### ISSUE TITLE
Launcher fails with iOS Simulator from Xcode 4.3 GM

### ISSUE THREAD

**Issue by chucks** [landonf/simlaunch#8]
At present, iOS apps bundled with simlaunch won't successfully launch if the user has Xcode 4.3 GM installed.  This appears to be for two separate reasons...both of which appear to be related to Apple moving the various Xcode-related tools such as iOS Simulator to be located directly inside the Xcode bundle.
-   PLSimulatorDiscovery searches for "iPhoneSimulator.platform" directories using Spotlight, and thus doesn't find the new platform directory that is inside the Xcode bundle.  I have tested a fix for this where if Spotlight fails to find the iPhone Simulator. platform directory, I'll also look directly where Xcode 4.3 would locate the file "/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform"

However, even if I fix this, there is a second problem. 
-  When PLSimulatorPlatform goes to load iPhoneSimulatorRemoteClient.framework, the load will fail because it can't find a dependant private framework.  The framework exists, but it has been moved to a different location, and is also inside Xcode's bundle.

console log:

```
2/16/12 2:24:07.963 PM Launcher: Failed to load iPhoneSimulatorRemoteClient framework: Error Domain=NSCocoaErrorDomain Code=3587 "The bundle “iPhoneSimulatorRemoteClient” couldn’t be loaded because it is damaged or missing necessary resources." (dlopen_preflight(/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/PrivateFrameworks/iPhoneSimulatorRemoteClient.framework/iPhoneSimulatorRemoteClient): Library not loaded: @rpath/DevToolsFoundation.framework/Versions/A/DevToolsFoundation
  Referenced from: /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/PrivateFrameworks/iPhoneSimulatorRemoteClient.framework/iPhoneSimulatorRemoteClient
  Reason: image not found) UserInfo=0x400454f00 {NSLocalizedFailureReason=The bundle is damaged or missing necessary resources., NSLocalizedRecoverySuggestion=Try reinstalling the bundle., NSFilePath=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/PrivateFrameworks/iPhoneSimulatorRemoteClient.framework/iPhoneSimulatorRemoteClient, NSDebugDescription=dlopen_preflight(/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/PrivateFrameworks/iPhoneSimulatorRemoteClient.framework/iPhoneSimulatorRemoteClient): Library not loaded: @rpath/DevToolsFoundation.framework/Versions/A/DevToolsFoundation
  Referenced from: /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/PrivateFrameworks/iPhoneSimulatorRemoteClient.framework/iPhoneSimulatorRemoteClient
  Reason: image not found, NSBundlePath=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/PrivateFrameworks/iPhoneSimulatorRemoteClient.framework, NSLocalizedDescription=The bundle “iPhoneSimulatorRemoteClient” couldn’t be loaded because it is damaged or missing necessary resources.}
```

In Xcode 4.2, DevToolsFoundation.framework was located at:

```
/Developer/Library/PrivateFrameworks
```

but in Xcode 4.3, it is located at:

```
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/PrivateFrameworks 
```

If I examine iPhoneSimulatorRemoteClient.framework in MacDependency, it also reports that it can't resolve the reference to DevToolsFoundation.framework.

Is there a problem with the Xcode 4.3 build of the private framework, or is there some build setting that could be tweaked in Launcher to search in additional locations for DevToolsFoundation?

**Comment by chucks** [comment]
If it's not clear from the console logs, the dylib that can't be resolved by iPhoneSimulatorRemoteClient is:

```
@rpath/DevToolsFoundation.framework/Versions/A/DevToolsFoundation
```

**Comment by ashcatch** [comment]
And the DevToolsFoundation.framework is located in /Applications/Xcode.app/Contents/OtherFrameworks -- see also https://jira.appcelerator.org/browse/TISTUD-1196 for a suggested fix.

**Comment by chucks** [comment]
Thanks for the pointer, I'll see if I can adapt this to simlaunch

**Comment by kimhunter** [comment]
If you run the md query and look for Xcode.app first then for the platform you can retrieve it, I'm working on a patch that will fallback to looking for the platform in Xcode. see below

⚡ mdfind "kMDItemDisplayName == 'iPhoneSimulator.platform'"

⚡ mdfind -onlyin /Applications/Xcode.app/Contents  "kMDItemDisplayName == 'iPhoneSimulator.platform'"
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform

**Comment by kimhunter** [comment]
I've got it working in Xcode 4.3 alongside 4.2(SL)

https://github.com/mogeneration/simlaunch/commits/standalone_xcode

Its not the best fix but its something i had to get working to get to the bottom of other problems.

**Comment by chucks** [comment]
I have slightly different fix for the same issue:  

https://github.com/chucks/simlaunch/commit/c96818b8e528e0e50d5cf28467e6c7064df9c7d9#diff-1

**Comment by kimhunter** [comment]
oh mine also fixes the DevToolsFoundation Framework thing (Edit: just realised you linked it in the project file, damm suppressed diff)

Chucks yours has a hardcoded path :(

Edit:
In my case i've got to support loading over many Xcode versions (nightmare)

**Comment by landonf** [comment]
Resolved through significant hackery in trunk. I implemented custom Mach-O parsing and pre-emptive resolution of @rpath-relative LC_LOAD_DYLIB library references. This allows the use of Xcode installed at any location on the drive, and should be backwards compatible with earlier Xcode releases.

This improves on some of the approaches given above, in that they set -rpath to a fixed path to /Applications/Xcode.app. Testing would be most appreciative.
